File: go/trees/104_maximum_depth_of_binary_tree.go
package main

import "fmt"

func main() {
	// Test cases

	// result: 3
	tree := createTree([]int{3, 9, 20, -1, -1, 15, 7}, 0)
	fmt.Println(maxDepth(tree))

	// result: 2
	tree = createTree([]int{1, -1, 2}, 0)
	fmt.Println(maxDepth(tree))
}

var depth int

func maxDepth(root *TreeNode) int {
	if root == nil {
		return depth
	}
	totalDepth := dft(root, 1)
	depth = 0
	return totalDepth
}

func dft(root *TreeNode, level int) int {
	if level > depth {
		depth = level
	}
	if root.Left != nil {
		dft(root.Left, level+1)
	}
	if root.Right != nil {
		dft(root.Right, level+1)
	}

	return depth
}

File: ruby/trees/104_maximum_depth_of_binary_tree.rb
# Definition for a binary tree node.
class TreeNode
  attr_accessor :val, :left, :right
  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

def build_tree(values, level = 0)
  return if values[level].nil?

  root = TreeNode.new(values[level])
  if level < values.size
    root.left = build_tree(values, 2 * level + 1)
    root.right = build_tree(values, 2 * level + 2)
  end
  root
end

@depth = 0
def dft(root, level = 1)
  @depth = level if level > @depth

  dft(root.left, level + 1) unless root.left.nil?
  dft(root.right, level + 1) unless root.right.nil?

  @depth
end

# @param {TreeNode} root
# @return {Integer}
def max_depth(root)
  @depth if root.nil?

  total_depth = dft(root)
  @depth = 0
  total_depth
end

def test
  # Test cases

  # result: 3
  puts max_depth(build_tree([3, 9, 20, nil, nil, 15, 7]))

  # result: 2
  puts max_depth(build_tree([1, nil, 2]))
end

test
